.editorconfig

```diff
@@ -1,16 +1,79 @@
 root = true
 
-[*]
+
+## Repository Configurations
+############################
+[.{git*,editorconfig,*.yml}]
 charset = utf-8
 end_of_line = lf
 indent_style = space
+indent_size = unset
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[.gitmodules]
+indent_style = tab
+
+
+## Shell Scripts
+################
+[*.sh]
+charset = utf-8
+end_of_line = lf
+indent_style = tab
+indent_size = unset
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+
+## Make Files
+#############
+[Makefile]
+charset = utf-8
+end_of_line = lf
+indent_style = tab
 indent_size = 2
+trim_trailing_whitespace = true
 insert_final_newline = true
+
+
+## Haskell Sources
+##################
+[*.hs]
+charset = utf-8
+end_of_line = unset
+indent_style = space
+indent_size = unset
 trim_trailing_whitespace = true
+insert_final_newline = true
+
 
+## Markdown
+###########
 [*.{markdown,md}]
+charset = utf-8
+end_of_line = unset
+indent_style = space
+indent_size = unset
 trim_trailing_whitespace = false
```

Comment:
I personally never use trailing spaces for hard line-breaks, but we need to check with @jgm for this setting, after all GFM does support that feature so potentially it could be used in the GFM documentation or templates, and if I remember correctly pandoc also supports the feature via extensions.

I always set `trim_trailing_whitespace` to _true_ in my projects, because trailing space are the main cause of commits "noise", leading to hard-to-read diffs and (often) spurious contents changes.

Reply:
I'm sure there are tests that include trailing whitespace, even if we don't use it in documentation (and for all I recall, we might).

Reply:
I've revised the rules for `test/` folder in commit 6080aeb, so they apply to markdown files only — basically the only checks now are for UTF-8 encoding and EOL consistency, all other rules are dropped to allow testing edge-cases (including trailing spaces).

>  even if we don't use it in documentation (and for all I recall, we might).

I've checked. the only markdown file with trailing spaces is `COPYING.md`, which contains some trailing spaces that are just spurious and could actually be removed. 

So, @jgm, should we set the rules for markdown files outside the `test/` folder to forbid trailing WS? (and cleanup the  `COPYING.md` doc) 

Unless we need to use them in the documentation, it would prevent unclean commits.